Store separate WCPay Product and Price IDs and hashes metadata for test and live mode

mattallan opened the pull request:
Fixes #2943 

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->
 - Store WCPay Product/Price IDs and hashes in postmeta with `_test` and `_live` suffixes 

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

With our WCPay Subscriptions integration with Stripe Billing, when a product in WC is created/updated we store and update the following postmeta on products:
 - `_wcpay_product_id`
 - `_wcpay_product_price_id`
 - `_wcpay_product_hash`
 - `_wcpay_product_price_hash`

The product and price IDs are then passed to sever when attempting to a WCPay Subscriptions. There's an issue when a store switches from test mode to live mode of WC Payments because the `_wcpay_product_id` we created in test mode, doesn't exist in live mode.

This PR stores the above meta data with `_test` and `_live` so that we're storing all product and price data for both test/live environments.

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

I'm not able to flick my WC Pay dev site between live and test mode however these testing instructions assume you can.

1. Deactivate WC Subscriptions extension if you have it
2. Checkout `develop` branch
3. Enable WCPay Subscriptions from the WC Pay dev tools page
4. Set WC Payments gateway to test mode
4. Create a new Subscription product
5. Confirm the product is created on Stripes dashboard
5. Purchase this new subscription product
6. Switch your gateway to live mode (if you can)
7. Try to purchase this subscription product again
8. You should see an error on the checkout (this is caused by the `_wcpay_product_/price_id` not existing)
9. Checkout this branch `fix/issue-2943`
10. Try to purchase this subscription again
11. Checkout should work
12. You should also notice in your DB that product/price meta is stored with _test and _live (https://d.pr/i/RbacYN)

-------------------

Closes #2943

haszari commented:
> I was thinking about this some more and I realised there might be a potential issue if you switch between test mode between product changes.

Good points @james-allan  - thanks for thinking about this. 

The most important thing is making sure live mode data is correct/up to date. Keeping dev mode data in sync is less important IMO. Related question: what syncing between live/dev does Stripe platform do?

mattallan commented:
> I was thinking about this some more and I realised there might be a potential issue if you switch between test mode between product changes

Yes! Good catch!

I've just pushed up some changes to a lot of the Product Service class to work around this issue (warning: a lot has changed!)

Here's a summary of the latest:

- When a store manager updates a product in WC, we'll now try get any test and live mode wcpay product IDs and wcpay price IDs for this product, update both records to reflect the latest changes in WC.
- When we trash and untrash/restor a product in WC, we'll also now archive and unarchive both test and live products in Stripe.

In order to achieve this, we now have to pass around which environment (live or test) we're trying to get and/or update. While most of the changes should be pretty straight forward, there is a lot changing.

To test this, I've just been creating test products and updating prices and product details to make sure they're being pushed up to Stripe.

I haven't been able to test a situation where the store has both a live and test wcpay product Id that needs updating, however I did write a new unit test for this case.

---

@aprea are you able to flick your dev site between test and live mode? Because I'm connected to the main Stripe account, I'm not able to go live mode...

If you can test live mode, would you be able to quickly check these changes? 🙏 
Here's some testing instructions: 

1. While WC Payments is in live mode, create a new subscription product
2. Check the DB and you should only see a `_wcpay_product_id_live` postmeta on the product (no `_wcpay_product_id_test`)
3. Switch WC Payments to test mode, update the product to +$5 and maybe change the name
4. Check the DB, you should now also see `_wcpay_product_id_test` postmeta on the product
5. Check your stripe account to confirm that both the live and test product have the latest changes (+$5 and name is updated)

From my hacky testing this should all work but it would be nice to confirm this. thanks

aprea commented:
Please ignore my last comment, I spoke with Matt on Slack and the issue I reported is not actually an issue and is the expected behaviour.

When a product in WC has both `_wcpay_product_id_live` and `_wcpay_product_id_test` metadata, updating it **should** update the corresponding Stripe products in both test and live mode.

mattallan commented:
I've updated this PR on the latest `develop` branch and also updated the `exclude_meta_wcpay_product_id()`: https://github.com/Automattic/woocommerce-payments/pull/3100#discussion_r727715706

Merging this! Thanks @james-allan and @aprea for testing
